### PR TITLE
feat: Non auto height invisible widgets

### LINKED ISSUE #18681: [Bug]: TypeError: Cannot read properties of undefined (reading 'bottomRow')
Sentry Issue: [APPSMITH-5CS](https://sentry.io/organizations/appsmith/issues/3770898865/?referrer=github_integration)

```
TypeError: Cannot read properties of undefined (reading 'bottomRow')
  at getParentCurrentHeightInRows (sagas/autoHeightSagas/helpers.ts:63:43)
  at call (sagas/autoHeightSagas/widgets.ts:444:49)
...
(45 additional frame(s) were not displayed)
```

### LINKED ISSUE #19983: [Feature]: Collapse all invisible widgets in view and preview mode
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

- When a widget within a container -- which has auto height enabled -- goes invisible -- in view and preview mode -- they should fully collapse, allowing widgets beneath them to take up the vacated space.
- We cannot have this enabled by default without consequences to existing apps. The reason for this, is that this will effect exiting page layouts and cause unexpected behaviour without the knowledge of the application developer.
- To mitigate this, we should enable this only for newly created applications. This way, application developers will notice the behaviour during development.
- This requires the application API response, to provide the client with a flag that informs the client of the fact that the application is a newly created application, allowing the client to enable this feature without consequences like breaking apps.
- The application API response can have a flag `collapseInvisibleWidgets: true`.
- If this flag exists and is `true`, the client will collapse all invisible widgets within an auto height enabled container.
- If this flag doesn't exist or is `false`, the client will not collapse all invisible widgets within an auto height enabled container.

### Why should this be worked on?

- Today, widgets which have auto height enabled, will collapse, allowing widgets below them to occupy their space.
- However, not all widgets have auto height. This leads to widgets like Button and Divider, always taking up space.
- Collapsing invisible widgets is a basic expectation when it comes to web UI.

### PR DESCRIPTION
**Pull request by riodeuno**
## Description
This PR adds another feature update we had planned for Auto Height
- [ ] For new applications, in View and Preview mode, any widget which is invisible will let go of its space and collapse if it's either on the main Canvas or a container-like widget which has Auto-height enabled.
- [ ] Widgets within a container-like Widget, say Tabs, that doesn't have Auto-height enabled, will now let go of their space if they're invisible.
- [ ] The experience in Edit mode has not changed.

TL;DR: In new applications, in the Preview and Published _AKA_ View modes, if a widget is invisible and within an Auto-height-enabled container like a Tab, a Modal, a Form, or the main Canvas, it will fully collapse, allowing widgets below it to move up and take its space. This changes the behavior today prior to the release of this PR for Auto-height-enabled widgets.

Fixes #19983
Fixes #18681

#### Before

https://user-images.githubusercontent.com/103687/214884767-0e7b0984-329b-4c99-8cf2-38a8aecc51d9.mov

#### After

https://user-images.githubusercontent.com/103687/214884513-55345213-9402-4f2f-b588-ea6de3094de8.mov

## Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?
- Manual: See Test Plan for details
- Jest: New functions have jest tests
- Cypress: We check the positions of the widgets after trying to collapse a non-auto height widget within an auto height enabled container

### Test Plan
https://github.com/appsmithorg/TestSmith/issues/2176

### Issues raised during DP testing
https://github.com/appsmithorg/appsmith/pull/20118#issuecomment-1409937455
https://github.com/appsmithorg/appsmith/pull/20118#issuecomment-1410389053

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [NA] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [NA] PR is being merged under a feature flag

### QA activity:
- [x] Test plan has been approved by relevant developers
- [x] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith @Sripriya93 This is ready for a test plan and QA.
There is one cypress test pending. I'll have it ready by Monday.

**Comment by kamakshibhat-appsmith** [comment]
Hi @riodeuno need some clarifications
> TL;DR: In new applications, in the preview and published(view) mode, if a widget X, is invisible and within an auto height enabled container|Tab|Modal|Form|Main Canvas, the widget X will fully collapse, allowing widgets below it to move up and take up space. This is the behaviour which we have today (prior to the release of this PR) for auto height enabled widgets.

here by Main Canvas , Do you mean the widgets in the Main canvas will collapse if it has auto height enabled? 
And for rest container like widgets, all the child widgets will collapse (be it autoheight/fixed), if the parent has auto height enabled

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith Both points are correct. For main canvas, it should only happen only for newer applications. Could you check if this is the case in new applications?

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith Please use the latest DP. I've pushed an update. You were right in pointing out the issue with Point 1. There was a logical error which I missed somehow. I've fixed it.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { ObjectsRegistry } from "../../../../support/Objects/Registry";
```
Comment (riodeuno) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:1]: @Aishwarya-U-R @NandanAnantharamu Please review the cypress test.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+const { AggregateHelper, CommonLocators } = ObjectsRegistry;
+
+describe("Property Pane Code Commenting", () => {
```
Comment (riodeuno) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:5]: Describe this correctly.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  it("2. Divider should move up by the height of the button widget in view mode", () => {
+    AggregateHelper.AssertElementVisible(
+      CommonLocators._previewModeToggle("edit"),
+    );
+    AggregateHelper.GetNClick(CommonLocators._previewModeToggle("edit"));
+
+    cy.get(CommonLocators._dividerWidget).should("have.css", "top", "16px");
+    cy.get(CommonLocators._checkboxWidget).should("have.css", "top", "6px");
+  });
+});
```
Comment (NandanAnantharamu) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:33]: We can also add one test for publish mode as well.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+const { AggregateHelper, CommonLocators } = ObjectsRegistry;
+
+describe("Property Pane Code Commenting", () => {
+  before(() => {
+    cy.fixture("autoHeightInvisibleWidgetsDSL").then((val: any) => {
+      AggregateHelper.AddDsl(val);
+    });
+  });
+
```
Comment (NandanAnantharamu) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:11]: Tests description can be more descriptive

**Review by marks0351 — DISMISSED** [review]

**Comment by kamakshibhat-appsmith** [comment]
Issues:

**Comment by kamakshibhat-appsmith** [comment]
Issues:
1. Some sort of logic failing in main canvas because of which unusually large spaces are getting created between widgets when we toggle visibility [LOOM](https://www.loom.com/share/d9eac69b650d4822bd87a2c3644e7944) [APP](https://appsmith-git-feat-collapse-non-auto-height-ea92ed-get-appsmith.vercel.app/app/instant-updates-2/page2-63d8cd6e2d47b6174f566f5f/edit)

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno Issues PART -2 
6. The Buttons in the form widget is getting cut off when we toggle visibility [LOOM](https://www.loom.com/share/8e8b1d2151024811aaa8a0d78e618040)
7. when the modal's submit/close button are toggled for visibility, the gap between the button and the modal border increases by some length [LOOM](https://www.loom.com/share/5fb412eb15d2456c87ed87d6dae797a1) 
8. Another modal issue [LOOM](https://www.loom.com/share/71d9e3940d074d46b6ba8be84dd9bb08)

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts (hunk 3 above)
Comment (riodeuno) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:45]: Done!

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts (hunk 4 above)
Comment (riodeuno) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/DynamicHeight/DynamicHeight_Invisible_Widgets_spec.ts:13]: Updated description!

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith I have fixed all the issues reported. Please approve if all looks good to you.
Also, could you create a test plan link, so that I can add that here, and add the test plan label to this PR?

@NandanAnantharamu I've made the changes you requested. Please review.

**Review by rahulramesha — APPROVED** [review]

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno  Point 7and point 8 are still not fixed

**Review by rahulramesha — DISMISSED** [review]

**Comment by kamakshibhat-appsmith** [comment]
There came an issue where invisible containers were overlapping during switching of page. However, i am unable to replicate it now
Rest of the things look good @riodeuno

**Comment by kamakshibhat-appsmith** [comment]
Steps to replicate the overlapping issue : [LOOM](https://www.loom.com/share/c9f765bd01eb49f19671cfc59f528b12)
Link to [APP](https://appsmith-git-feat-collapse-non-auto-height-ea92ed-get-appsmith.vercel.app/app/overlapping-widget/send-messages-63dccd802b0e66264b2b37ee/edit#7bptx78qzc)

@riodeuno

File: app/client/src/sagas/PageSagas.tsx [hunk 5]
```diff
@@ -571,12 +572,14 @@ export function* saveLayoutSaga(action: ReduxAction<{ isRetry?: boolean }>) {
   try {
     const currentPageId: string = yield select(getCurrentPageId);
     const currentPage: Page = yield select(getPageById(currentPageId));
+    const isPreviewMode: boolean = yield select(previewModeSelector);
```
Comment (riodeuno) [inline app/client/src/sagas/PageSagas.tsx:575]: @jsartisan @keyurparalkar Please review this piece of code.

Comment (jsartisan) [reply app/client/src/sagas/PageSagas.tsx:575]: Looks good Abhinav!

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith We need to test the following:
- All Auto Height features
- Themeing
- Templates

File: app/client/src/utils/autoHeight/generateTree.ts [hunk 6]
```diff
@@ -2,45 +2,58 @@ import { areIntersecting } from "utils/boxHelpers";
...
+    const a: NodeSpace = spaceMap[A];
+    const b: NodeSpace = spaceMap[B];
     //if both are of the same level and previous tree exists, check originalTops
     if (a.top === b.top && previousTree[a.id] && previousTree[b.id]) {
+      if (
+        previousTree[a.id].originalTopRow -
+          previousTree[b.id].originalTopRow ===
+        0
+      ) {
+        return (
```
Comment (rahulramesha) [inline app/client/src/utils/autoHeight/generateTree.ts:27]: Not sure about this, this will make the one with the largest height be put on top, maybe even reversing the order on the layout. But I think this is still working because of preview mode fix in that scenario.

**Review by rahulramesha — APPROVED** [review]

File: app/client/src/utils/autoHeight/generateTree.ts (hunk 6 above)
Comment (riodeuno) [reply app/client/src/utils/autoHeight/generateTree.ts:27]: If the height was large, the bottomrow will be higher for the bottom widget. I see what you mean, I need to do a reverse of this where I do a B - A instead of A - B.

Comment (rahulramesha) [reply app/client/src/utils/autoHeight/generateTree.ts:27]: Not really. What i meant was, in the preview scenario we encountered, topRow of both the widgets will be 0 and whichever widget has the most height will end up on top because it will have higher bottomRow, Regardless of which one was at the top originally.

Comment (riodeuno) [reply app/client/src/utils/autoHeight/generateTree.ts:27]: You're right. In that case, we have to rely on the preview mode not saving layouts. I'll remove the bottomRow check.

**Comment by Sripriya93** [comment]
When we preview template from "Start from template" option in canvas. For the first template we are seeing the selected template (Eg. Customer support dashboard), now if we close the template and "Start from template" and select any other template other than the previously selected one (Eg. slack bot), we are seeing the one we selected first in this case "Customer support dashboard), for the third one it starts from the beginning. @riodeuno, CC. @kamakshibhat-appsmith 

https://user-images.githubusercontent.com/104895602/217807179-b96bd899-2490-454f-97af-86c938df6ea1.mp4

**Comment by riodeuno** [comment]
@Sripriya93 Is this not happening on release?

**Comment by Sripriya93** [comment]
No @riodeuno, its not happening in release or production and also not from the home page, but also happens when we preview it from the "Add page from template"

image

**Comment by Sripriya93** [comment]
@riodeuno Im able to reproduce this but its inconsistent with the steps, sometimes the issue happens when we close the Modal else it happens when we choose <- Back to templates option, below loom is after i refreshed and tried in both options

https://user-images.githubusercontent.com/104895602/217811549-c4432d6a-3ad6-4b4d-b416-7076f591444c.mp4

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno Pls check this issuehttps://www.loom.com/share/2892867da4c149e69f460fb6036a29b0

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno I couldn't help but notice we have Undo/redo/Cut/Copy/Paste allowed in Preview mode ; which if used will hamper the "removing Save Api call" fix  as the updates won't be synced. Let me know if this needs to be addressed at priority

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith Preview mode is just for previewing what the applications would look like in view mode. This means that preview mode should not change the layout in anyway. undo/redo/copy/paste are ideally not supposed to persist from the preview mode. So, we're in fact, making preview mode predictable by disabling the syncing of the page updates.

**Comment by kamakshibhat-appsmith** [comment]
Issue : There seems to be scroll coming up in deploy mode when a child widget inside container-type widgets leaves its space due to invisibility [LOOM](https://www.loom.com/share/ffea640931bd419b9846a8806279c503)

**Comment by kamakshibhat-appsmith** [comment]
Issue : There seems to be scroll coming up in deploy mode when a child widget inside container-type widgets leaves its space due to invisibility [LOOM](https://www.loom.com/share/ffea640931bd419b9846a8806279c503) @riodeuno

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith Thank you for finding this issue! Yes, this was due to an error in the code. I've fixed it.

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith @Sripriya93 This PR is ready from the CI front. Only the flaky tests are failing. Please QA this once more, and if all looks good. We'll merge this into release.

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno  We can replicate the same scenario in deploy mode when there is just one child widget inside container type widget and the child is set to invisible . [LOOM](https://www.loom.com/share/c0efe7ec11d047d480dbba48111d365a)

> Issue : There seems to be scroll coming up in deploy mode when a child widget inside container-type widgets leaves its space due to invisibility . We seem to have regressed [LOOM](https://www.loom.com/share/ffea640931bd419b9846a8806279c503)

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith This is actually a separate edge case. Once again, really appreciate the fact that you found this, makes the release more stable. I'll fix this as soon as I can.

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith I've resolved both the following issues:
https://github.com/appsmithorg/appsmith/pull/20118#issuecomment-1427104093
https://github.com/appsmithorg/appsmith/pull/19082#issuecomment-1427108999

Please verify.

**Comment by kamakshibhat-appsmith** [comment]
@riodeuno Noticing new issues with auto height computation for invisible widgets [LOOM](https://www.loom.com/share/d3c593470708488281ae11466db91e55)

**Comment by riodeuno** [comment]
@kamakshibhat-appsmith This a great catch! I've pushed a fix, it should be available in the DP in 15 minutes or so.

**Review by marks0351 — APPROVED** [review]

**Review by rahulramesha — APPROVED** [review]

**Comment by kamakshibhat-appsmith** [comment]
LGTM 👍  , We can proceed  @riodeuno
